s3-pusher does not work if `~/.aws/*` files are missing

## Case 1
-  `GO_SERVER_DASHBOARD_URL`, `GO_ARTIFACTS_S3_BUCKET`, `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION` set to correct values in my gocd environment
- No `~/.aws/config` and `~/.aws/credentials` file present
Result:

```
[go] Task: Plugin with ID: indix.s3publishtook: 7.777s
Error: Interaction with plugin with id 'indix.s3publish' implementing 'task' extension failed while requesting for 'execute'. Reason: [Unable to find a region via the region provider chain. Must provide an explicit region in the builder or setup environment to supply a region.]
```

Relevant stacktrace:
```
Caused by: com.amazonaws.SdkClientException: Unable to find a region via the region provider chain. Must provide an explicit region in the builder or setup environment to supply a region.
        at com.amazonaws.client.builder.AwsClientBuilder.setRegion(AwsClientBuilder.java:371)
        at com.amazonaws.client.builder.AwsClientBuilder.configureMutableProperties(AwsClientBuilder.java:337)
        at com.amazonaws.client.builder.AwsSyncClientBuilder.build(AwsSyncClientBuilder.java:46)
        at com.indix.gocd.utils.store.S3ArtifactStore.getS3client(S3ArtifactStore.java:209)
        at com.indix.gocd.utils.store.S3ArtifactStore.<init>(S3ArtifactStore.java:45)
        at com.indix.gocd.s3publish.PublishExecutor.getS3ArtifactStore(PublishExecutor.java:89)
        at com.indix.gocd.s3publish.PublishExecutor.execute(PublishExecutor.java:45)
        at com.indix.gocd.s3publish.PublishTask.handleTaskExecution(PublishTask.java:74)
        at com.indix.gocd.s3publish.PublishTask.handle(PublishTask.java:46)
        at com.thoughtworks.go.plugin.infra.DefaultPluginManager$2.execute(DefaultPluginManager.java:166)
        at com.thoughtworks.go.plugin.infra.DefaultPluginManager$2.execute(DefaultPluginManager.java:161)
        at com.thoughtworks.go.plugin.infra.FelixGoPluginOSGiFramework.executeActionOnTheService(FelixGoPluginOSGiFramework.java:315)
```

## Case 2
- Environment variables set like in Case 1
- File `~/.aws/config` contains:
   ```
   [default]
   region = eu-central-1
   ```
Result:

```
Error: Interaction with plugin with id 'indix.s3publish' implementing 'task' extension failed while requesting for 'execute'. Reason: [null]
```

Relevant stacktrace: (unfortunately it's missing the most interesting part because it's cut-off:

```
java.lang.RuntimeException: Interaction with plugin with id 'indix.s3publish' implementing 'task' extension failed while requesting for 'execute'. Reason: [null]
        at com.thoughtworks.go.plugin.access.PluginRequestHelper.submitRequest(PluginRequestHelper.java:41)
        at com.thoughtworks.go.plugin.access.pluggabletask.JsonBasedTaskExecutor.execute(JsonBasedTaskExecutor.java:41)
        at com.thoughtworks.go.domain.builder.pluggableTask.PluggableTaskBuilder.executeTask(PluggableTaskBuilder.java:97)
        at com.thoughtworks.go.domain.builder.pluggableTask.PluggableTaskBuilder$1.execute(PluggableTaskBuilder.java:74)
        at com.thoughtworks.go.domain.builder.pluggableTask.PluggableTaskBuilder$1.execute(PluggableTaskBuilder.java:71)
        at com.thoughtworks.go.plugin.access.pluggabletask.TaskExtension.execute(TaskExtension.java:57)
        at com.thoughtworks.go.domain.builder.pluggableTask.PluggableTaskBuilder.build(PluggableTaskBuilder.java:71)
        at com.thoughtworks.go.remote.work.Builders.build(Builders.java:75)
        at com.thoughtworks.go.remote.work.BuildWork.execute(BuildWork.java:241)
        at com.thoughtworks.go.remote.work.BuildWork.buildJob(BuildWork.java:206)
        at com.thoughtworks.go.remote.work.BuildWork.build(BuildWork.java:148)
        at com.thoughtworks.go.remote.work.BuildWork.doWork(BuildWork.java:93)
        at com.thoughtworks.go.agent.JobRunner.run(JobRunner.java:60)
        at com.thoughtworks.go.agent.AgentHTTPClientController.retrieveWork(AgentHTTPClientController.java:123)
        at com.thoughtworks.go.agent.AgentHTTPClientController.work(AgentHTTPClientController.java:99)
        at com.thoughtworks.go.agent.AgentController.loop(AgentController.java:78)
        at sun.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.util.MethodInvoker.invoke(MethodInvoker.java:273)
        at org.springframework.scheduling.support.MethodInvokingRunnable.run(MethodInvokingRunnable.java:65)
        at org.springframework.scheduling.timer.DelegatingTimerTask.run(DelegatingTimerTask.java:70)
        at java.util.TimerThread.mainLoop(Timer.java:555)
        at java.util.TimerThread.run(Timer.java:505)
Caused by: java.lang.RuntimeException
        at com.thoughtworks.go.plugin.infra.FelixGoPluginOSGiFramework.executeActionOnTheService(FelixGoPluginOSGiFramework.java:317)
        at com.thoughtworks.go.plugin.infra.FelixGoPluginOSGiFramework.doOn(FelixGoPluginOSGiFramework.java:245)
        at com.thoughtworks.go.plugin.infra.DefaultPluginManager.submitTo(DefaultPluginManager.java:161)
        at com.thoughtworks.go.plugin.access.PluginRequestHelper.submitRequest(PluginRequestHelper.java:32)
        ... 23 more
Caused by: java.lang.StackOverflowError
        at com.google.gson.stream.JsonWriter.writeDeferredName(JsonWriter.java:402)
        at com.google.gson.stream.JsonWriter.value(JsonWriter.java:417)
        at com.google.gson.internal.bind.TypeAdapters$13.write(TypeAdapters.java:362)
        at com.google.gson.internal.bind.TypeAdapters$13.write(TypeAdapters.java:346)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:68)
        at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.write(MapTypeAdapterFactory.java:208)
        at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.write(MapTypeAdapterFactory.java:145)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:68)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:89)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.write(ReflectiveTypeAdapterFactory.java:195)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:68)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:89)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.write(ReflectiveTypeAdapterFactory.java:195)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:68)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:89)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.write(ReflectiveTypeAdapterFactory.java:195)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:68)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:89)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.write(ReflectiveTypeAdapterFactory.java:195)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:68)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:89)
...
```

## Case 3
- Environment variables and `~/.aws/config` like in Case 2
- `~/.aws/credentials` contains:
   ```
   [default]
   aws_access_key_id = XXX
   aws_secret_access_key = ZZZ   
   ```
Result: now the upload completes successfully

## Case 4
- Same setting as in Case 3, however change environment variables `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` to invalid values

Result: upload still completes successfully

So, to summarize:
* If i just set the environment variables it doesn't work
* If `config` and `credentials` in `~/.aws` is missing it doesn't work
* If `~/.aws` files are present and correct the environment variables have no effect at all

--- a/build.sbt
+++ b/build.sbt
@@ -12,7 +12,7 @@ val junitInterface = "com.novocode" % "junit-interface" % "0.11" % Test
 val mockito = "org.mockito" % "mockito-all" % "1.10.19" % Test
 val scalaTest = "org.scalatest" %% "scalatest" % "2.2.0" % Test
 
-val appVersion = sys.env.get("TRAVIS_TAG") orElse sys.env.get("BUILD_LABEL") getOrElse "1.0.0-SNAPSHOT"
+val appVersion = sys.env.get("TRAVIS_TAG") orElse sys.env.get("BUILD_LABEL") getOrElse s"1.0.0-${System.currentTimeMillis / 1000}-SNAPSHOT"
 
 lazy val root = Project(
   id = "gocd-s3-artifacts",

--- a/fetch/src/main/java/com/indix/gocd/s3fetch/FetchExecutor.java
+++ b/fetch/src/main/java/com/indix/gocd/s3fetch/FetchExecutor.java
@@ -20,10 +20,6 @@ public class FetchExecutor {
 
     public TaskExecutionResult execute(Config config, final Context context) {
         final GoEnvironment env = new GoEnvironment(context.getEnvironmentVariables());
-        if (!env.hasAWSUseIamRole()) {
-            if (env.isAbsent(AWS_ACCESS_KEY_ID)) return envNotFound(AWS_ACCESS_KEY_ID);
-            if (env.isAbsent(AWS_SECRET_ACCESS_KEY)) return envNotFound(AWS_SECRET_ACCESS_KEY);
-        }
         if (env.isAbsent(GO_ARTIFACTS_S3_BUCKET)) return envNotFound(GO_ARTIFACTS_S3_BUCKET);
 
         try {

--- a/fetch/src/main/resources/views/task.template.html
+++ b/fetch/src/main/resources/views/task.template.html
@@ -14,5 +14,5 @@
   <span class="form_error" ng-show="GOINPUTNAME[Destination].$error.server">{{ GOINPUTNAME[Destination].$error.server }}</span>
 </div>
 <div class="form_item_block">
-    <p class="required">Make sure either AWS_USE_IAM_ROLE or AWS_SECRET_ACCESS_KEY + AWS_ACCESS_KEY_ID and GO_ARTIFACTS_S3_BUCKET environment variables are present with appropriate values on any of pipeline / Go Environments / on all agent machines. </p>
+    <p class="required">Make sure GO_ARTIFACTS_S3_BUCKET environment variables is present with appropriate values on any of pipeline / Go Environments / on all agent machines. </p>
 </div>

--- a/fetch/src/test/java/com/indix/gocd/s3fetch/FetchExecutorTest.java
+++ b/fetch/src/test/java/com/indix/gocd/s3fetch/FetchExecutorTest.java
@@ -51,15 +51,6 @@ public class FetchExecutorTest {
     }
 
     @Test
-    public void shouldBeFailureIfEnvVarNotFound() {
-        Map<String, String> mockVariables = mockEnvironmentVariables.with(AWS_ACCESS_KEY_ID, "").build();
-        TaskExecutionResult result = fetchExecutor.execute(config, mockContext(mockVariables));
-
-        assertFalse(result.isSuccessful());
-        assertThat(result.message(), is("AWS_ACCESS_KEY_ID environment variable not present"));
-    }
-
-    @Test
     public void shouldBeFailureIfFetchConfigNotValid() {
         Map<String, String> mockVariables = mockEnvironmentVariables.build();
         config = new Config(Maps.builder()

--- a/publish/src/main/java/com/indix/gocd/s3publish/PublishExecutor.java
+++ b/publish/src/main/java/com/indix/gocd/s3publish/PublishExecutor.java
@@ -34,10 +34,6 @@ public class PublishExecutor {
 
     public TaskExecutionResult execute(Config config, final Context context) {
         final GoEnvironment env = new GoEnvironment(context.getEnvironmentVariables());
-        if (!env.hasAWSUseIamRole()) {
-            if (env.isAbsent(AWS_ACCESS_KEY_ID)) return envNotFound(AWS_ACCESS_KEY_ID);
-            if (env.isAbsent(AWS_SECRET_ACCESS_KEY)) return envNotFound(AWS_SECRET_ACCESS_KEY);
-        }
         if (env.isAbsent(GO_ARTIFACTS_S3_BUCKET)) return envNotFound(GO_ARTIFACTS_S3_BUCKET);
         if (env.isAbsent(GO_SERVER_DASHBOARD_URL)) return envNotFound(GO_SERVER_DASHBOARD_URL);
 
@@ -49,24 +45,23 @@ public class PublishExecutor {
 
         try {
             List<SourceDestination> sourceDestinations = config.sourceDestinations();
-            foreach(sourceDestinations, new VoidFunction<SourceDestination>() {
-                @Override
-                public void execute(final SourceDestination input) {
-                    String[] files = parseSourcePath(input.source, context.getWorkingDir());
+            for(SourceDestination input : sourceDestinations) {
+                String[] files = parseSourcePath(input.source, context.getWorkingDir());
+                for (String includedFile : files) {
+                    File localFileToUpload = new File(String.format("%s/%s", context.getWorkingDir(), includedFile));
+                    if (!fileExists(localFileToUpload)) {
+                        throw new RuntimeException(String.format("%s is missing", localFileToUpload.getAbsolutePath()));
+                    }
 
-                    foreach(files, new VoidFunction<String>() {
-                        @Override
-                        public void execute(String includedFile) {
-                            File localFileToUpload = new File(String.format("%s/%s", context.getWorkingDir(), includedFile));
-                            if (!fileExists(localFileToUpload)) {
-                                throw new RuntimeException(String.format("%s is missing", localFileToUpload.getAbsolutePath()));
-                            }
-
-                            pushToS3(context, destinationPrefix, store, localFileToUpload, input.destination);
-                        }
-                    });
+                    pushToS3(context, destinationPrefix, store, localFileToUpload, input.destination);
                 }
-            });
+            }
+
+            if(!hasConfigDestinationPrefix(config)) {
+                setMetadata(env, bucket, destinationPrefix, store);
+            }
+
+            return new TaskExecutionResult(true, "Published all artifacts to S3");
         } catch (JsonSyntaxException e) {
             String message = "Failed while parsing configuration";
             logger.error(message);
@@ -75,14 +70,6 @@ public class PublishExecutor {
             logger.error(e.getMessage(), e);
             return new TaskExecutionResult(false, e.getMessage(), e);
         }
-
-        // A configured destination prefix is used to deploy files rather than publish artifacts
-        // We only want to set metadata when publishing artifacts
-        if(!hasConfigDestinationPrefix(config)) {
-            setMetadata(env, bucket, destinationPrefix, store);
-        }
-
-        return new TaskExecutionResult(true,"Published all artifacts to S3");
     }
 
     protected S3ArtifactStore getS3ArtifactStore(GoEnvironment env, String bucket) {

--- a/publish/src/main/java/com/indix/gocd/s3publish/PublishExecutor.java
+++ b/publish/src/main/java/com/indix/gocd/s3publish/PublishExecutor.java
@@ -1,8 +1,6 @@
 package com.indix.gocd.s3publish;
 
-import com.amazonaws.auth.InstanceProfileCredentialsProvider;
-import com.amazonaws.services.s3.AmazonS3;
-import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
 import com.amazonaws.services.s3.model.ObjectMetadata;
 import com.amazonaws.services.s3.model.PutObjectRequest;
 import com.google.gson.JsonSyntaxException;
@@ -33,26 +31,25 @@ public class PublishExecutor {
     private Logger logger = Logger.getLoggerFor(PublishTask.class);
 
     public TaskExecutionResult execute(Config config, final Context context) {
-        final GoEnvironment env = new GoEnvironment(context.getEnvironmentVariables());
-        if (env.isAbsent(GO_ARTIFACTS_S3_BUCKET)) return envNotFound(GO_ARTIFACTS_S3_BUCKET);
-        if (env.isAbsent(GO_SERVER_DASHBOARD_URL)) return envNotFound(GO_SERVER_DASHBOARD_URL);
-
-        final String bucket = env.get(GO_ARTIFACTS_S3_BUCKET);
-        final S3ArtifactStore store = getS3ArtifactStore(env, bucket);
-        store.setStorageClass(env.getOrElse(AWS_STORAGE_CLASS, STORAGE_CLASS_STANDARD));
-
-        final String destinationPrefix = getDestinationPrefix(config, env);
-
         try {
+            final GoEnvironment env = new GoEnvironment(context.getEnvironmentVariables());
+            if (env.isAbsent(GO_ARTIFACTS_S3_BUCKET)) return envNotFound(GO_ARTIFACTS_S3_BUCKET);
+            if (env.isAbsent(GO_SERVER_DASHBOARD_URL)) return envNotFound(GO_SERVER_DASHBOARD_URL);
+
+            final String bucket = env.get(GO_ARTIFACTS_S3_BUCKET);
+            final S3ArtifactStore store = getS3ArtifactStore(env, bucket);
+            store.setStorageClass(env.getOrElse(AWS_STORAGE_CLASS, STORAGE_CLASS_STANDARD));
+
+            final String destinationPrefix = getDestinationPrefix(config, env);
+
             List<SourceDestination> sourceDestinations = config.sourceDestinations();
             for(SourceDestination input : sourceDestinations) {
                 String[] files = parseSourcePath(input.source, context.getWorkingDir());
+                if (files.length == 0) {
+                    return new TaskExecutionResult(false, String.format("Source %s didn't yield any files to upload", input.source));
+                }
                 for (String includedFile : files) {
                     File localFileToUpload = new File(String.format("%s/%s", context.getWorkingDir(), includedFile));
-                    if (!fileExists(localFileToUpload)) {
-                        throw new RuntimeException(String.format("%s is missing", localFileToUpload.getAbsolutePath()));
-                    }
-
                     pushToS3(context, destinationPrefix, store, localFileToUpload, input.destination);
                 }
             }
@@ -61,14 +58,14 @@ public class PublishExecutor {
                 setMetadata(env, bucket, destinationPrefix, store);
             }
 
-            return new TaskExecutionResult(true, "Published all artifacts to S3");
+            return new TaskExecutionResult(true, "Published all artifacts to S3 successfully");
         } catch (JsonSyntaxException e) {
             String message = "Failed while parsing configuration";
             logger.error(message);
             return new TaskExecutionResult(false, message, e);
         } catch (Exception e) {
             logger.error(e.getMessage(), e);
-            return new TaskExecutionResult(false, e.getMessage(), e);
+            return new TaskExecutionResult(false, e.getMessage());
         }
     }
 
@@ -133,7 +130,7 @@ public class PublishExecutor {
     }
 
     private TaskExecutionResult envNotFound(String environmentVariable) {
-        String message = String.format("%s environment variable not present", environmentVariable);
+        String message = String.format("%s environment variable is not set", environmentVariable);
         logger.error(message);
         return new TaskExecutionResult(false, message);
     }

--- a/publish/src/main/resources/views/task.template.html
+++ b/publish/src/main/resources/views/task.template.html
@@ -37,7 +37,7 @@
     </p>
 </div>
 <div class="form_item_block">
-    <p class="required">Make sure either AWS_USE_IAM_ROLE or AWS_SECRET_ACCESS_KEY + AWS_ACCESS_KEY_ID, GO_ARTIFACTS_S3_BUCKET and GO_SERVER_DASHBOARD_URL environment variables are present with appropriate values on any of pipeline / Go Environments / on all agent machines. </p>
+    <p class="required">Make sure GO_ARTIFACTS_S3_BUCKET and GO_SERVER_DASHBOARD_URL environment variables are present with appropriate values on any of pipeline / Go Environments / on all agent machines. </p>
 </div>
 <script type="text/javascript">
     jQuery(document).ready(function() {

--- a/publish/src/test/java/com/indix/gocd/s3publish/PublishExecutorTest.java
+++ b/publish/src/test/java/com/indix/gocd/s3publish/PublishExecutorTest.java
@@ -54,48 +54,6 @@ public class PublishExecutorTest {
     }
 
     @Test
-    public void shouldThrowIfAWS_ACCESS_KEY_IDNotPresent() {
-        Map<String, String> mockVariables = mockEnvironmentVariables.with(AWS_ACCESS_KEY_ID, "").build();
-
-        TaskExecutionResult result = publishExecutor.execute(config, mockContext(mockVariables));
-        assertFalse(result.isSuccessful());
-        assertThat(result.message(), is("AWS_ACCESS_KEY_ID environment variable not present"));
-    }
-
-    @Test
-    public void shouldThrowIfAWS_SECRET_ACCESS_KEYNotPresent() {
-        Map<String, String> mockVariables = mockEnvironmentVariables.with(AWS_SECRET_ACCESS_KEY, "").build();
-
-        TaskExecutionResult result = publishExecutor.execute(config, mockContext(mockVariables));
-        assertFalse(result.isSuccessful());
-        assertThat(result.message(), is("AWS_SECRET_ACCESS_KEY environment variable not present"));
-    }
-
-    @Test
-    public void shouldNotThrowIfAWSUseIAMRoleIsFalseAndAWS_ACCESS_KEY_IDNotPresent() {
-        Map<String, String> mockVariables = mockEnvironmentVariables
-                .with(AWS_USE_IAM_ROLE, "False")
-                .with(AWS_ACCESS_KEY_ID, "")
-                .build();
-
-        TaskExecutionResult result = publishExecutor.execute(config, mockContext(mockVariables));
-        assertFalse(result.isSuccessful());
-        assertThat(result.message(), is("AWS_ACCESS_KEY_ID environment variable not present"));
-    }
-
-    @Test
-    public void shouldThrowIfAWSUseIAMRoleIsFalseAndAWS_SECRET_ACCESS_KEYNotPresent() {
-        Map<String, String> mockVariables = mockEnvironmentVariables
-                .with(AWS_USE_IAM_ROLE, "False")
-                .with(AWS_SECRET_ACCESS_KEY, "")
-                .build();
-
-        TaskExecutionResult result = publishExecutor.execute(config, mockContext(mockVariables));
-        assertFalse(result.isSuccessful());
-        assertThat(result.message(), is("AWS_SECRET_ACCESS_KEY environment variable not present"));
-    }
-
-    @Test
     public void shouldThrowIfLocalFileDoesNotExist() {
         doReturn(false).when(publishExecutor).fileExists(any(File.class));
         AmazonS3Client mockClient = mockClient();

--- a/publish/src/test/java/com/indix/gocd/s3publish/PublishExecutorTest.java
+++ b/publish/src/test/java/com/indix/gocd/s3publish/PublishExecutorTest.java
@@ -54,8 +54,7 @@ public class PublishExecutorTest {
     }
 
     @Test
-    public void shouldThrowIfLocalFileDoesNotExist() {
-        doReturn(false).when(publishExecutor).fileExists(any(File.class));
+    public void shouldFailIfNoFilesToUploadBasedOnSource() {
         AmazonS3Client mockClient = mockClient();
 
         Config config = new Config(Maps.builder()
@@ -66,11 +65,11 @@ public class PublishExecutorTest {
         TaskExecutionResult result = executeMockPublish(
                 mockClient,
                 config,
-                new String[]{"README.md"}
+                new String[]{}
         );
 
         assertFalse(result.isSuccessful());
-        assertThat(result.message(), containsString("README.md is missing"));
+        assertThat(result.message(), containsString("Source target/* didn't yield any files to upload"));
     }
 
     @Test
@@ -103,7 +102,7 @@ public class PublishExecutorTest {
 
         TaskExecutionResult result = publishExecutor.execute(config, mockContext(mockVariables));
         assertFalse(result.isSuccessful());
-        assertThat(result.message(), is("GO_ARTIFACTS_S3_BUCKET environment variable not present"));
+        assertThat(result.message(), is("GO_ARTIFACTS_S3_BUCKET environment variable is not set"));
     }
 
     @Test
@@ -122,7 +121,7 @@ public class PublishExecutorTest {
         );
 
         assertTrue(result.isSuccessful());
-        assertThat(result.message(), is("Published all artifacts to S3"));
+        assertThat(result.message(), is("Published all artifacts to S3 successfully"));
     }
 
     @Test

--- a/utils/src/main/java/com/indix/gocd/utils/Constants.java
+++ b/utils/src/main/java/com/indix/gocd/utils/Constants.java
@@ -13,6 +13,7 @@ public class Constants {
 
     public static final String AWS_SECRET_ACCESS_KEY = "AWS_SECRET_ACCESS_KEY";
     public static final String AWS_ACCESS_KEY_ID = "AWS_ACCESS_KEY_ID";
+    public static final String AWS_REGION = "AWS_REGION";
     public static final String AWS_USE_IAM_ROLE = "AWS_USE_IAM_ROLE";
     public static final String AWS_STORAGE_CLASS = "AWS_STORAGE_CLASS";
     public static final String STORAGE_CLASS_STANDARD = "standard";

--- a/utils/src/main/java/com/indix/gocd/utils/TaskExecutionResult.java
+++ b/utils/src/main/java/com/indix/gocd/utils/TaskExecutionResult.java
@@ -24,7 +24,6 @@ public class TaskExecutionResult {
         final HashMap result = new HashMap();
         result.put("success", success);
         result.put("message", message);
-        result.put("exception", exception);
         return result;
     }
 

--- a/utils/src/main/java/com/indix/gocd/utils/store/S3ArtifactStore.java
+++ b/utils/src/main/java/com/indix/gocd/utils/store/S3ArtifactStore.java
@@ -1,5 +1,7 @@
 package com.indix.gocd.utils.store;
 
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
 import com.amazonaws.auth.InstanceProfileCredentialsProvider;
 import com.amazonaws.services.s3.AmazonS3;
 import com.amazonaws.services.s3.AmazonS3Client;
@@ -202,9 +204,17 @@ public class S3ArtifactStore {
 
     public static AmazonS3 getS3client(GoEnvironment env) {
         AmazonS3ClientBuilder amazonS3ClientBuilder = AmazonS3ClientBuilder.standard();
+
+        if(env.has(AWS_REGION)) {
+            amazonS3ClientBuilder.withRegion(env.get(AWS_REGION));
+        }
         if (env.hasAWSUseIamRole()) {
             amazonS3ClientBuilder.withCredentials(new InstanceProfileCredentialsProvider(false));
+        } else if (env.has(AWS_ACCESS_KEY_ID) && env.has(AWS_SECRET_ACCESS_KEY)) {
+            BasicAWSCredentials basicCreds = new BasicAWSCredentials(env.get(AWS_ACCESS_KEY_ID), env.get(AWS_SECRET_ACCESS_KEY));
+            amazonS3ClientBuilder.withCredentials(new AWSStaticCredentialsProvider(basicCreds));
         }
+
 
         return amazonS3ClientBuilder.build();
     }

